iothub/device/src/Transport/Amqp/AmqpConnectionHolder.cs

```diff
@@ -183,7 +183,7 @@ public async Task<AmqpIotConnection> EnsureConnectionAsync(CancellationToken can
             }
             catch (OperationCanceledException)
             {
-                throw new TimeoutException();
+                throw new IotHubClientException(true, IotHubStatusCode.NetworkErrors);
```

Comment:
If the operation was cancelled, why would we label this as a network error?

Reply:
For this particular case, the service may just be very busy and is too slow to respond to the connect request even if the underlying network is fine

Reply:
I am using `IotHubStatusCode.NetworkErrors` to replace `IotHubCommunicationException` (which we had before) which was thrown when an attempt to communicate with the IoT hub service fails due to transient network errors after exhausting all the retries based on the retry policy set on the client or due to operation timeouts. As the original exception thrown here was `TimeoutException`, I think `NetworkErrors` fits the place here

Reply:
> I am using `IotHubStatusCode.NetworkErrors` to replace `IotHubCommunicationException` (which we had [before](https://github.com/Azure/azure-iot-sdk-csharp/blob/fe800fcbbadb627db689e4d01566f03bde1ae9ba/iothub/device/src/Exceptions/IotHubCommunicationException.cs)) which was thrown when an attempt to communicate with the IoT hub service fails due to transient network errors after exhausting all the retries based on the retry policy set on the client or due to operation timeouts. 

I'm fine with this, sure

> As the original exception thrown here was `TimeoutException`, I think `NetworkErrors` fits the place here

This doesn't follow from the above to me. I'd rather we have an IotHubStatusCode for ```Timeout``` for cases like these where the SDK has no way of knowing if the request failed because of network issues or because the service was too slow

Reply:
I considered about using a new `IotHubStatusCode.Timeout` before, but then I thought `NetworkErrors` would cover the case of operation timeouts.
Additional question for handling the case of " service was too slow": we do have `IotHubStatusCode.ServerBusy` now, if using a new code `Timeout` here, will it be sort of overlapping?

Reply:
The point of a ```Timeout``` enum value would be to signal that the SDK does not know if it was caused by network issues or by service being slow

> Additional question for handling the case of " service was too slow": we do have IotHubStatusCode.ServerBusy now, if using a new code Timeout here, will it be sort of overlapping?

We only use ```ServerBusy``` when the service explicitly notifies the SDK that it is busy. Timeout would be for cases when the SDK doesn't receive any sort of notification from the service

Reply:
Or, to put it another way, if a user sees the SDK emitting ```NetworkErrors``` reasons for failure when the customer has validated that their network to the hub is fine, they will likely file a bug.

Reply:
that's fair!